File: majority-element-ii/majority-element-ii.java
class Solution {
    public List<Integer> majorityElement(int[] nums) {
        int m1 = 0 , m2 = 0, cnt1 = 0 , cnt2 = 0;
        for(int i=0;i<nums.length;i++)
        {
            if(nums[i] == m1) cnt1++;
            else if(nums[i] == m2) cnt2++;
            else if(cnt1 == 0) {cnt1 = 1 ; m1 = nums[i];}
            else if(cnt2 == 0) {cnt2 = 1; m2 = nums[i];}
            else {cnt1--; cnt2--;}
        }
        
        cnt1 = 0; cnt2 = 0;
        for(int i=0;i<nums.length;i++)
        {
            if(nums[i] == m1) cnt1++;
            else if(nums[i] == m2) cnt2++;
        }
        
        List<Integer> ans = new ArrayList<Integer>();
        
        if(cnt1 > nums.length/3) ans.add(m1);
        if(cnt2 > nums.length/3) ans.add(m2);
        
        return ans;
    }
}

File: majority-element-ii/majority-element-ii.cpp
class Solution {
    bool check(vector<int> &arr,int can)
    {
        int n = arr.size();
        int cnt = 0;
        for(int i=0;i<n;i++) cnt += arr[i] == can;

        return cnt > n/3;
    }

public:
    vector<int> majorityElement(vector<int>& A) {
        int val1= INT_MIN , val2= INT_MIN, cnt1=0, cnt2=0;
        int n = A.size();

        for(int i=0;i<n;i++)
        {
            if(A[i] == val1) cnt1++;
            else if(A[i] == val2) cnt2++;
            else if(cnt1 == 0)
            {
                val1 = A[i]; cnt1++;
            }
            else if(cnt2 == 0)
            {
                val2 = A[i]; cnt2++;
            }
            else 
            {
                cnt1--; cnt2--;
            }
        }
        
        vector<int> res;
        
        if(check(A,val1)) res.push_back(val1);
        if(check(A,val2)) res.push_back(val2);
        
        return res;
    }
};